### PR TITLE
first draft of CONTRIBUTING.MD

### PR DESCRIPTION
**Pull request by McKnight-42**
resolves #https://github.com/dbt-labs/core-team/issues/33

### Description

A first draft of the possible CONTRIUBTING.MD file for the Adapter Repos

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-bigquery next" section.

### CODE REVIEW

**Comment by McKnight-42** [comment]
Would love any advice on format, if information seems clear, if we need all the same sexctions from the dbt-core Contrubting.md file etc. once we get this one done should be just a template for all the others which will make finishing assignment quick so we can take very discerning eyes with this one.

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:89]: A lot of this applies to `CONTRIBUTING.md` in `dbt-core` as well
```suggestion
### dbt Labs contributors
```

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:53]: All of the places that currently say `dbt` can be `dbt-bigquery`. (Makes me realize that it should probably say `dbt-core` [over there](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md).) Let's anticipate a world where there are actually different maintainers for `dbt-core` + `dbt-bigquery`

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:71]:
```suggestion
- **Trunks** are where active development of the next release takes place. There is one trunk named `main` at the time of writing this, and will be the default branch of the repository.
```

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:128]:
```suggestion
First make sure that you set up your `virtualenv` as described in [Setting up an environment](#setting-up-an-environment). Also ensure you have the latest version of pip installed with `pip install --upgrade pip`.

Next, perform an ["editable"](https://pip.pypa.io/en/stable/cli/pip_install/#editable-installs) install of `dbt-bigquery`, plus some extra dependencies used in testing:
```

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:131]:
```suggestion
pip install -e . -r dev-requirements.txt
```
Note that **this is a place where our adapter plugins will differ from dbt-core**. In the `dbt-core` repo, this needs to be:
```
pip install -r editable-requirements.txt -r dev-requirements.txt
```
And in dbt-spark (special case), this needs to include [extras](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) like so:
```
pip install -e .[all] -r dev-requirements.txt
```

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:134]: Let's also encourage them to run commands like:
```
$ which dbt
$ dbt --version
```
to confirm they're using the right one

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:100]: No more py36:
```suggestion
- [`tox`](https://tox.readthedocs.io/en/latest/) to manage virtualenvs across python versions. We currently target the latest patch releases for Python 3.7, Python 3.8, and Python 3.9
```

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:142]: - remove mention of py36 (true across the board)
- we don't currently run `mypy` checks on adapter plugins (note: **this is a place where adapter plugins differ from dbt-core**)

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:156]: It's worth going into a bit more detail here:
- docs are hosted here: https://docs.getdbt.com/
- all the code is defined in a repo here: https://github.com/dbt-labs/docs.getdbt.com
- (**specific to this plugin**) in particular, contributions to this repo are likely to impact one/both of: [BigQuery profile](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile), [BigQuery configs](https://docs.getdbt.com/reference/resource-configs/bigquery-configs)
- we ask every community contributor who's made a user-facing change to open an issue for their change in that repo
    - in fact, let's add a checklist item to the [PR template in this repo](https://github.com/dbt-labs/dbt-bigquery/blob/main/.github/pull_request_template.md) + in other adapter plugin repos
    - @runleonarun: we need better issue templates in the docs.getdbt.com repo! I'm thinking questions like: which docs pages need to change? what's the big idea / context behind the change? any sample code / examples you can share?
- if the contributor feels up to it, they should contribute the PR for the docs change, too!

File: CONTRIBUTING.md
Comment (jtcohen6) [inline CONTRIBUTING.md:161]: We can be less vague here, since we know their PR is in the `dbt-bigquery` repo and wants to run BigQuery integration tests.

That said, we should say things like:
- If you're a first-time contributor, code checks + unit tests require a maintainer to approve
- A maintainer will need to add the `ok to test` label, at which point your subsequent commits will trigger integration tests in BigQuery

I believe those are true across dbt-core + all adapter plugins. We can decide if it's worth reiterating here, for folks who skim the core contributing guide

**Review by jtcohen6 — COMMENTED** [review]
@McKnight-42 Thanks for pulling this together! Reading through this has really helped me think through what we want out of these contributing guides, and what ought to live in each adapter plugin vs. in the "core" contributing guide.

I see there's a _lot_ of duplication with the info in the `dbt-core` contributing doc. I'm thinking we should:
- Update the `dbt-core` contributing doc! A lot of the feedback I left below is relevant there, too
- At the top of the `dbt-bigquery` contributing doc, say something like: `The dbt-bigquery plugin is maintained by dbt Labs. We recommend that you read the [core contributing doc](https://github.com/dbt-labs/dbt-core/blob/HEAD/CONTRIBUTING.md) first, if you haven't already—most of the information there is quite useful for contributing here, too!` This will prevent us from needing to coordinate updates across many repos

Then, the `dbt-bigquery` contributing doc can include _just_ the pieces that are **different** from `dbt-core`, which I've bolded in my comments below. The big ones are:
- Cloning + installing `dbt-bigquery` locally: use `pip install -e .`
- Running integration tests locally: let's detail the `BIGQUERY_TEST_*` env vars that are required
- Which docs pages are likely to be impacted by user-facing changes
- Refer to the repo/codebase/package/maintainers as `dbt-bigquery` throughout

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md:53]: I didn't know if we ever purposefully used say like a over arching general reference to dbt as a whole so tried to go off context but I completely agree.

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:83]:
````suggestion
## Testing
### Initial Setup

`dbt-bigquery` uses test credentials specified in a `test.env` file in the root of the repository. This `test.env` file is git-ignored, but please be _extra_ careful to never check in credentials or other sensitive information when developing. To create your `test.env` file, copy the provided example file, then supply your relevant credentials.

```
cp test.env.example test.env
$EDITOR test.env
```

````

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:86]:
```suggestion
There are a few methods for running tests locally.
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:88]:
```suggestion
#### `tox`
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:89]:
```suggestion
`tox` takes care of managing Python virtualenvs and installing dependencies in order to run tests. You can also run tests in parallel, for example you can run unit tests for Python 3.7, Python 3.8, Python 3.9, and `flake8` checks in parallel with `tox -p`. Also, you can run unit tests for specific python versions with `tox -e py37`. The configuration of these tests are located in `tox.ini`.
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:91]:
```suggestion
#### `pytest`
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:92]:
```suggestion
Finally, you can also run a specific test or group of tests using `pytest` directly. With a Python virtualenv active and dev dependencies installed you can do things like:
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:109]: I think you can remove these lines. Or suggest creating an issue or a PR for docs changes.

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:114]:
```suggestion
dbt Labs provides a CI environment to test changes to the `dbt-bigquery` adapter, and periodic checks against the development version of `dbt-core` through Github Actions. 
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:4]: just a reminder to remove comment before merging

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:47]:
```suggestion
First make sure that you set up your `virtualenv` as described in [Setting up an environment](https://github.com/dbt-labs/dbt-core/blob/HEAD/CONTRIBUTING.md#setting-up-an-environment).  Ensure you have the latest version of pip installed with `pip install --upgrade pip`. Next, install `dbt-bigquery` latest dependencies:
```

**Review by kwigley — CHANGES_REQUESTED** [review]
looking good, here are some initial suggestions!

File: CONTRIBUTING.md
Comment (leahwicz) [inline CONTRIBUTING.md:51]:
```suggestion
To confirm you have the correct version of `dbt-core` installed please run `dbt --version` and `which dbt`.
```

File: CONTRIBUTING.md
Comment (leahwicz) [inline CONTRIBUTING.md:95]:
```suggestion
dbt Labs provides a CI environment to test changes to the `dbt-bigquery` adapter and periodic checks against the development version of `dbt-core` through Github Actions.  
```

File: CONTRIBUTING.md
Comment (leahwicz) [inline CONTRIBUTING.md:98]: Can we add a little blurb here about needing to add your change to the CHANGELOG?

File: .gitignore
Comment (leahwicz) [inline .gitignore:53]: Nit: is there a reason for this line addition? If not, lets get rid of it

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md:51]: change made locally will push up

File: .gitignore
Comment (McKnight-42) [reply .gitignore:53]: change made locally will push up

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md:97]: added a small section to the paragraph on line 97

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md]: changes made locally and pushed up

File: CONTRIBUTING.md
Comment (ChenyuLInx) [inline CONTRIBUTING.md:51]: Is there something we can run to also confirm that we are using the right dbt-bigquery?

File: CONTRIBUTING.md
Comment (ChenyuLInx) [inline CONTRIBUTING.md:22]: Do most folks using mac do this to get git? I know it works but seems a bit extra thing that needs to go in. Just nitpicking here

File: CONTRIBUTING.md
Comment (ChenyuLInx) [inline CONTRIBUTING.md:57]: Maybe a link for more context around this section?

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md:51]: the best option i know of would be pip list and checking the version of the dbt-adaptername that way doesn't seem like --version or which work for them.

Comment (ChenyuLInx) [reply CONTRIBUTING.md:51]: I feel like `pip freeze | grep bigquery` should do the trick

Comment (McKnight-42) [reply CONTRIBUTING.md:51]: For me it shows the bigquery version but only the github hashed value of the current branch for dbt-bigquery which could be used to compare that your enviornment is using the branch you want if that doesn't seem to obtuse but I really like the suggestion.

File: CONTRIBUTING.md
Comment (McKnight-42) [reply CONTRIBUTING.md:57]: @kwigley do you know of any relevant links for this? I was looking through the docs but haven't found anything specific yet.

File: CONTRIBUTING.md
Comment (ChenyuLInx) [reply CONTRIBUTING.md:51]: Sounds good! And I realized that pip would only work for folks installed dbt via pip. And `dbt --version` would actually show all adaptor version

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:13]:
```suggestion
For those wishing to contribute we highly suggest reading the [dbt-core](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md), if you haven't already. Almost all of the information there is applicable to contributing here, too!
```

File: CONTRIBUTING.md
Comment (kwigley) [reply CONTRIBUTING.md:22]:
```suggestion
You will need `git` in order to download and modify the `dbt-bigquery` source code. You can find direction [here](https://github.com/git-guides/install-git) on how to install `git`.
```

File: CONTRIBUTING.md
Comment (kwigley) [reply CONTRIBUTING.md:57]:
```suggestion
`dbt-bigquery` contains [unit](https://github.com/dbt-labs/dbt-bigquery/tree/main/tests/unit) and [integration](https://github.com/dbt-labs/dbt-bigquery/tree/main/tests/integration) tests. Integration tests require testing against an actual BigQuery warehouse. We have CI set up to test against a BigQuery warehouse. In order to run integration tests locally, you will need a `test.env` file in the root of the repository that contains credentials for BigQuery.

Note: This `test.env` file is git-ignored, but please be _extra_ careful to never check in credentials or other sensitive information when developing. To create your `test.env` file, copy the provided example file, then supply your relevant credentials.
```

File: CONTRIBUTING.md
Comment (kwigley) [inline CONTRIBUTING.md:70]:
```suggestion
#### `pytest`
```

**Review by kwigley — CHANGES_REQUESTED** [review]
👍 looks good, included some tweaks

**Review by kwigley — APPROVED** [review]

**Pull request merged**
